### PR TITLE
Introduced WindowedJoinOperatorFeatures classes to solve the problem …

### PR DESCRIPTION
**Pull request by davidyan74**
…of code duplication

@ShunxinLu In this PR, I'm trying to minimize code duplication. Please take a look and provide any feedback.

### CODE REVIEW

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+  public void processWatermark2(ControlTuple.Watermark watermark)
+  {
+    latestWatermark2 = watermark.getTimestamp();
+    long minWatermark = Math.min(latestWatermark1, latestWatermark2);
+    operator.setCurrentWatermark(minWatermark);
+  }
+
+  private void setCurrentWatermark(long timestamp)
+  {
```
Comment (ShunxinLu) [inline library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java]: Why this method is taking in a timestamp but not using it?

**Comment by davidyan74** [comment]
Also, note here that you won't have to duplicate the code that supports session window because `KeyedWindowedJoinOperatorImpl` now extends  `KeyedWindowedOperatorImpl`

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    public Keyed(AbstractWindowedOperator<KeyValPair<KeyT, InputT1>, ?, DataStorageT, ?, AccumulationT> operator)
+    {
+      super(operator);
+    }
+
+    protected <T extends KeyValPair> void accumulateTuple(Tuple.WindowedTuple<T> tuple, AccumFunction<T> accumFn)
+    {
+      final KeyValPair<KeyT, T> kvData = tuple.getValue();
```
Comment (ShunxinLu) [inline library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java:146]: Why the value of the KeyValPair here is another KeyValPair?

**Comment by ShunxinLu** [comment]
@davidyan74 I am still digesting the implementation of `WindowedJoinOperatorFeatures`. But from what I understand it certainly can eliminate most of the duplicate code. Thank you so much for your help!

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java (hunk 2 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java:146]: Note that the getValue is on the tuple, which is a wrapper of the tuple value. The tuple value is a KeyValPair.

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java (hunk 1 above)
Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java]: Good catch. I'm removing this method

**Review by davidyan74 — COMMENTED** [review]
Please review the changes again and incorporate them if you think they are okay. Also, please add unit tests for join operators when you have time. We can merge the main PR once that's done.

File: library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java (hunk 2 above)
Comment (ShunxinLu) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java:146]: Sorry I didn't explain clearly. Here the kvData is the value of the tuple which is a KeyValPair, this should have no issue. But the type of the value of kvData is T, which extends KeyValPair. What I meant in my previous question is why the value of kvData is another KeyValPair?

Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java:146]: You're absolutely correct. I have pushed the fix. Please take a look again.

Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java:146]: That's why we need unit tests for join :)

**Comment by ShunxinLu** [comment]
@davidyan74 Thanks David! Everything looks good to me. Should I merge this PR to my branch and continue to work on it?

**Comment by davidyan74** [comment]
@ShunxinLu Yes, please do so. I think the only thing that is missing now is proper unit tests for the join operator.

**Pull request merged**